### PR TITLE
added SnowflakeIO with Read operation

### PR DESCRIPTION
**Pull request by DariuszAniszewski**
Internal review

### CODE REVIEW

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakePipelineOptions.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  @Description("Temporary GCS bucket name")
+  String getStagingBucketName();
+
+  void setStagingBucketName(String stagingBucketName);
+
+  @Description("Internal (local) location name to connect to.")
+  String getInternalLocation();
+
+  void setInternalLocation(String internalLocation);
+
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakePipelineOptions.java]: As we don't support internal I think we can remove it

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+ * There are also other options available to configure connection to Snowflake:
+ *
+ * <ul>
+ *   <li>{@link DataSourceConfiguration#withWarehouse(String)} to specify which Warehouse to use
+ *   <li>{@link DataSourceConfiguration#withDatabase(String)} to specify which Database to connect
+ *       to
+ *   <li>{@link DataSourceConfiguration#withSchema(String)} to specify which schema to use
+ *   <li>{@link DataSourceConfiguration#withRole(String)} to specify which role to use
+ *   <li>{@link DataSourceConfiguration#withPortNumber(int)} to specify custom port of Snowflake
+ *       instance
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:89]: In `DataSourceConfiguration` there are even more options, should we add them? 
withAuthenticator
withLoginTimeout

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+
+  /**
+   * Read data from Snowflake via COPY statement via user-defined {@link SnowflakeService}.
+   *
+   * @param snowflakeService user-defined {@link SnowflakeService}
+   * @param <T> Type of the data to be read.
+   */
+  public static <T> Read<T> read(
+      SnowflakeService snowflakeService, SnowflakeCloudProvider snowFlakeCloudProvider) {
+    return new AutoValue_SnowflakeIO_Read.Builder<T>()
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:137]: Sometimes we use `snowFlakeCloudProvider ` or `cloudProvider` as a variable name for `SnowflakeCloudProvider`. Maybe is worth to make consistent names in the whole codebase ?

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+      checkArgument(
+          (getDataSourceProviderFn() != null),
+          "withDataSourceConfiguration() or withDataSourceProviderFn() is required");
+
+      String gcpTmpDirName = makeTmpDirName();
+      PCollection<T> output;
+      PCollection<Void> emptyCollection = input.apply(Create.of((Void) null));
+
+      output =
+          emptyCollection
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:320]: I think we can have here like this :
`PCollection<T> output; = emptyCollection.appply .....`

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+    public abstract ValueProvider<String> getRole();
+
+    @Nullable
+    public abstract ValueProvider<String> getAuthenticator();
+
+    @Nullable
+    public abstract ValueProvider<Integer> getLoginTimeout();
+
+    @Nullable
+    public abstract ValueProvider<Boolean> getSsl();
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: In Read and Write methods we stopped using `ValueProvider ` maybe here we also don't need it ?

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/credentials/SnowflakeCredentialsFactory.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+public class SnowflakeCredentialsFactory {
+  public static SnowflakeCredentials of(SnowflakePipelineOptions options) {
+    if (options.getOauthToken() != null && !options.getOauthToken().isEmpty()) {
+      return new OAuthTokenSnowflakeCredentials(options.getOauthToken());
+    } else if (options.getUsername() != null
+        && !options.getUsername().isEmpty()
+        && !options.getPassword().isEmpty()) {
+      return new UsernamePasswordSnowflakeCredentials(options.getUsername(), options.getPassword());
+    } else if (options.getUsername() != null
+        && !options.getUsername().isEmpty()
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/credentials/SnowflakeCredentialsFactory.java]: Maybe is worth to wrap those checks in separate methods ?

**Review by purbanow — APPROVED** [review]
Looks good to me :)

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakePipelineOptions.java (hunk 1 above)
Comment (DariuszAniszewski) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakePipelineOptions.java]: 👌

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 2 above)
Comment (DariuszAniszewski) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:89]: `authenticator` is fixed when proper `SnowflakeCredentials` are used. I will remove it
`loginTimetout` - we can add it here

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 3 above)
Comment (DariuszAniszewski) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:137]: I made it `snowflakeCloudProvider` all over the place

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 5 above)
Comment (DariuszAniszewski) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: 💯

**Pull request closed without merging**